### PR TITLE
Establish seed node connections in async during node bootstrap

### PR DESCRIPTION
**Pull request by ankitkala**
### Description
Establish seed node connections in async during node bootstrap

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7950

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by krishna-ggk — COMMENTED** [review]
So one consequence of this could be that node is up while remote connection establishment is still in progress in which case any outbound cross-cluster requests could fail with "no such remote" right?

Could this also impact CCR?

**Comment by shwetathareja** [comment]
@ankitkala can you please highlight the implication of moving the remote connection to async. As i understand today also there is no guarantee that connection would be established in sync in 10secs. What happens in case a CCR/ CCS request comes. What is the error thrown?

**Comment by ankitkala** [comment]
> So one consequence of this could be that node is up while remote connection establishment is still in progress in which case any outbound cross-cluster requests could fail with "no such remote" right?
> 
> Could this also impact CCR?

> @ankitkala can you please highlight the implication of moving the remote connection to async. As i understand today also there is no guarantee that connection would be established in sync in 10secs. What happens in case a CCR/ CCS request comes. What is the error thrown?


This should not impact any of the cross cluster calls. For each such call, Remote cluster client invokes a blocking call to `remoteClusterService.ensureConnected` to ensure that the connection is established [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/RemoteClusterAwareClient.java#L69) ([ensureConnected](https://github.com/opensearch-project/OpenSearch/blob/0a2009250cd857d68cdfa87340f8cd1b906d8c6e/server/src/main/java/org/opensearch/transport/RemoteClusterConnection.java#L117) and [connectImpl](https://github.com/opensearch-project/OpenSearch/blob/1e08b5a075fa6018be6c0af959b2d638c2743d56/server/src/main/java/org/opensearch/transport/SniffConnectionStrategy.java#L244))
This code change makes the connect as async only during node bootstrap. Cross cluster calls and Seed node updates aren't affected by this change.

**Comment by ankitkala** [comment]
> The one common problem with async connection handling is the race around connection and disconnection. See https://github.com/opensearch-project/OpenSearch/issues/4874. Want to ensure we don't run into a similar issue going forward

Thanks. I took a cursory look at the issue above. 
This change isn't modifying the behaviour of how the connection is established/disconnected. So I don't expect any issues there.

File: server/src/main/java/org/opensearch/transport/RemoteClusterService.java [hunk 1]
```diff
@@ -328,35 +326,35 @@ synchronized void updateRemoteCluster(String clusterAlias, Settings newSettings,
...
+        GroupedActionListener<Void> groupListener = new GroupedActionListener<>(
+            ActionListener.wrap(r -> listener.onResponse(null), listener::onFailure),
+            enabledClusters.size()
+        );
 
-        if (enabledClusters.isEmpty()) {
-            future.onResponse(null);
-        }
+        for (String clusterAlias : enabledClusters) {
+            updateRemoteCluster(clusterAlias, settings, new ActionListener<>() {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/transport/RemoteClusterService.java:345]: do we need another wrapper ActionListener, we can add logs in the listener created for GroupListener in line 340

Comment (ankitkala) [reply server/src/main/java/org/opensearch/transport/RemoteClusterService.java:345]: Yeah, not need technically. 
Group listener will log aggregate failure across multiple aliases which can be confusing. So i thought of adding explicit logs per alias/remote cluster. 
Let me know if you think if it should be removed.

Comment (ankitkala) [reply server/src/main/java/org/opensearch/transport/RemoteClusterService.java]: Updated.

**Review by shwetathareja — APPROVED** [review]

**Pull request merged**